File: src/main/java/mediathek/mainwindow/FilmAgeLabel.java
package mediathek.mainwindow;

import mediathek.config.Daten;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.time.Duration;
import java.util.IllegalFormatException;

public class FilmAgeLabel extends JLabel implements ActionListener {
    record FilmListAge(long hours, long minutes) {
    }

    private FilmListAge calculateFilmListAge() {
        var duration = Duration.ofSeconds(Daten.getInstance().getListeFilme().getMetaData().getAgeInSeconds());
        var minutes = duration.toMinutes();
        var hours = minutes / 60;
        minutes -= hours * 60;
        return new FilmListAge(hours, minutes);
    }

    private final Timer timer;
    private FilmListAge oldAge = new FilmListAge(0, 0);

    public FilmAgeLabel() {
        setToolTipText("Alter der Filmliste");

        setAgeToLabel();

        timer = new Timer(1000, this);
        timer.setRepeats(true);
        timer.start();
    }

    private String computeAgeString(@NotNull FilmListAge age) throws IllegalFormatException {
        if (age.hours == 0) {
            return String.format("Alter: %dm", age.minutes);
        } else {
            return String.format("Alter: %dh %dm", age.hours, age.minutes);
        }
    }

    private void setAgeToLabel() {
        var curAge = calculateFilmListAge();
        if (!curAge.equals(oldAge)) {
            var result = computeAgeString(curAge);
            setText(result);
            oldAge = curAge;
        }
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        setAgeToLabel();
    }
}

File: src/main/kotlin/mediathek/mainwindow/FilmAgeLabel.kt
package mediathek.mainwindow

import mediathek.config.Daten
import org.apache.commons.lang3.time.DurationFormatUtils
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.time.Duration
import java.util.*
import java.util.concurrent.TimeUnit
import javax.swing.JLabel
import javax.swing.Timer

class FilmAgeLabel : JLabel(), ActionListener {
    private data class FilmListAge(val hours: Long, val minutes: Long)

    private var oldAge = FilmListAge(0, 0)

    init {
        toolTipText = "Alter der Filmliste"
        setAgeToLabel()

        //start the update timer
        val timer = Timer(1000, this)
        timer.isRepeats = true
        timer.start()
    }

    private fun calculateFilmListAge(): FilmListAge {
        val duration = Duration.ofSeconds(Daten.getInstance().listeFilme.metaData.ageInSeconds)
        var minutes = duration.toMinutes()
        val hours = minutes / 60
        minutes -= hours * 60
        return FilmListAge(hours, minutes)
    }

    @Throws(IllegalFormatException::class)
    private fun computeAgeString(age: FilmListAge): String {
        return if (age.hours == 0L) {
            "Alter: ${age.minutes}m"
        } else if (age.hours >= 24) {
            val duration = TimeUnit.MILLISECONDS.convert(age.hours * 60 + age.minutes, TimeUnit.MINUTES)
            DurationFormatUtils.formatDuration(duration, "'Alter: 'dd'd' HH'h' mm'm'", true)
        } else {
            "Alter: ${age.hours}h ${age.minutes}m"
        }
    }

    private fun setAgeToLabel() {
        val curAge = calculateFilmListAge()
        if (curAge != oldAge) {
            val result = computeAgeString(curAge)
            text = result
            oldAge = curAge
        }
    }

    override fun actionPerformed(e: ActionEvent) {
        setAgeToLabel()
    }
}